### PR TITLE
Don't project _id field in Mongo DB connector when it's not required

### PR DESCRIPTION
**Pull request by sahoss**
## Description

Currently, _id field is retrieved even when it's not necessary. We can exclude by https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/builders/projections/#exclusion-of-_id.

Fixes https://github.com/trinodb/trino/issues/17970

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 1]
```diff
@@ -142,6 +143,8 @@ public class MongoSession
     public static final String COLLECTION_NAME = "collectionName";
     public static final String ID = "id";
 
+    private static final String _ID = "_id";
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:146]: Please inline. Extracting this wouldn't improve the readability.

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoSession.java [hunk 2]
```diff
@@ -43,6 +43,34 @@ public class TestMongoSession
     private static final MongoColumnHandle COL3 = new MongoColumnHandle("col3", createUnboundedVarcharType(), false, Optional.empty());
     private static final MongoColumnHandle COL4 = new MongoColumnHandle("col4", BOOLEAN, false, Optional.empty());
 
+    private static final MongoColumnHandle _ID = new MongoColumnHandle("_id", ObjectIdType.OBJECT_ID, false, Optional.empty());
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoSession.java:46]: Please rename to `ID_COL`

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoSession.java [hunk 3]
```diff
@@ -43,6 +43,34 @@ public class TestMongoSession
...
+
+    @Test
+    public void testBuildProjectionWithoutId()
+    {
+        ImmutableList<MongoColumnHandle> columns = ImmutableList.of(COL1, COL2);
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoSession.java:51]:
```suggestion
        List<MongoColumnHandle> columns = ImmutableList.of(COL1, COL2);
```

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoSession.java [hunk 4]
```diff
@@ -43,6 +43,34 @@ public class TestMongoSession
...
+                .append(COL1.getName(), 1)
+                .append(COL2.getName(), 1)
+                .append(_ID.getName(), 0);
+        assertEquals(expected, output);
+    }
+
+    @Test
+    public void testBuildProjectionWithId()
+    {
+        ImmutableList<MongoColumnHandle> columns = ImmutableList.of(COL1, COL2, _ID);
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoSession.java:64]:
```suggestion
        List<MongoColumnHandle> columns = ImmutableList.of(COL1, COL2, _ID);
```

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoSession.java [hunk 5]
```diff
@@ -43,6 +43,34 @@ public class TestMongoSession
...
+    public void testBuildProjectionWithoutId()
+    {
+        ImmutableList<MongoColumnHandle> columns = ImmutableList.of(COL1, COL2);
+
+        Document output = MongoSession.buildProjection(columns);
+        Document expected = new Document()
+                .append(COL1.getName(), 1)
+                .append(COL2.getName(), 1)
+                .append(_ID.getName(), 0);
+        assertEquals(expected, output);
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoSession.java:58]: `expected` argument should be the 2nd argument.

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoSession.java [hunk 6]
```diff
@@ -43,6 +43,34 @@ public class TestMongoSession
...
+    public void testBuildProjectionWithId()
+    {
+        ImmutableList<MongoColumnHandle> columns = ImmutableList.of(COL1, COL2, _ID);
+
+        Document output = MongoSession.buildProjection(columns);
+        Document expected = new Document()
+                .append(COL1.getName(), 1)
+                .append(COL2.getName(), 1)
+                .append(_ID.getName(), 1);
+        assertEquals(expected, output);
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoSession.java:71]: `expected` argument should be the 2nd argument.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 7]
```diff
@@ -499,6 +499,25 @@ public MongoCursor<Document> execute(MongoTableHandle tableHandle, List<MongoCol
         return iterable.iterator();
     }
 
+    static Document buildProjection(List<MongoColumnHandle> columns)
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:502]: Add `@VisibleForTesting` annotation.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 8]
```diff
@@ -499,6 +499,25 @@ public MongoCursor<Document> execute(MongoTableHandle tableHandle, List<MongoCol
...
+    {
+        Document output = new Document();
+        // _id is always projected by mongodb unless its explicitly excluded.
+        // Explicitly exclude it, if it is not part of the list of columns
+        // https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/builders/projections/#exclusion-of-_id
+        boolean excludeId = true;
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:508]: I think we don't need this flag. We can put `output.append("_id", 0)` before `for` loop.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 9]
```diff
@@ -482,15 +483,12 @@ public long deleteDocuments(RemoteTableName remoteTableName, TupleDomain<ColumnH
...
-            output.append(column.getName(), 1);
-        }
+        Document projection = buildProjection(columns);
         MongoCollection<Document> collection = getCollection(tableHandle.getRemoteTableName());
         Document filter = buildFilter(tableHandle);
-        FindIterable<Document> iterable = collection.find(filter).projection(output).collation(SIMPLE_COLLATION);
+        FindIterable<Document> iterable = collection.find(filter).projection(projection).collation(SIMPLE_COLLATION);
         tableHandle.getLimit().ifPresent(iterable::limit);
-        log.debug("Find documents: collection: %s, filter: %s, projection: %s", tableHandle.getSchemaTableName(), filter, output);
+        log.info("Find documents: collection: %s, filter: %s, projection: %s", tableHandle.getSchemaTableName(), filter, projection);
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:491]: Please revert the unrelated change.

**Comment by sahoss** [comment]
@ebyhr addressed comments. please take a look.

**Comment by sahoss** [comment]
the tests which are failing are hive/hadoop related. 99% sure not related to this.

**Comment by Praveen2112** [comment]
Can we use Builder from MongoDB driver - https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/builders/#using-builders which has some predefined APIs we could use.

**Comment by sahoss** [comment]
> Can we use Builder from MongoDB driver - https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/builders/#using-builders which has some predefined APIs we could use.

Would it be better to do it in a different PR? as then the PR would be changing two things 1. excludeId 2. changing implementation of how projections are built. and neither are currently directly tested. @Praveen2112

**Comment by ebyhr** [comment]
Could you rebase on master to resolve conflicts?

**Review by ebyhr — APPROVED** [review]
